### ISSUE TITLE
BUG- file_download control doesn't allow dynamic files to be downloaded

### ISSUE THREAD

**Issue by Dr-Irv** [Avaiga/taipy-gui#916]
**Description**
The `file_download` control is broken.  What we want to do is have someone click the button, then the action is called to create the file, and then that file is downloaded, with the specified name.

What happens is that the file that is downloaded is the first one that was there, and then the action function is called.

Also, the `name` field seems to be ignored.

Ideally, what we would like is to create a temporary file in the file system and then after it is downloaded, the file is deleted. But we could use a `io.BytesIO` object as well, and just pass that, but if we do that the file is called `TaipyContent.bin`

**How to reproduce**

```python
import datetime as dt
from taipy import Gui
import pandas as pd

df = pd.DataFrame({"x": [1, 2, 3], "y": [4, 5, 6]})
filename = "foobar.xlsx"
df.to_excel(filename)


def action(state):
    tfilename = "tmp-" + dt.datetime.now().strftime("%H-%M-%S") + ".xlsx"
    df.to_excel(tfilename)
    state.filename = tfilename
    return


page = "<|{filename}|file_download|name=junk.xlsx|on_action=action|>"

if __name__ == "__main__":
    Gui(page=page).run()
```

**Expected behavior**
What should happen here is that the file that is created dynamically (which has a name based on the hour/minute/second the action function is called) is the one that is downloaded. But what happens is that the file that is downloaded is the *previous* filename, not the newly created one.

In addition, the name of `junk.xlsx` is not respected at all.

**Runtime environment**
Please specify relevant indications.
 - OS:  Windows 10 and Linux
 - Browser: Chrome

taipy 2.4.0

**Comment by FlorianJacta** [comment]
Thank you for the issue, Irv! Indeed, it will be ideal to have the _on_action_ callback called before downloading the file. Having it called after is also useful to delete the file for example. We will look into that!

**Comment by FredLL-Avaiga** [comment]
I suppose the easiest way to do this would be to have a standard button with an action  that would invoke [download(state, ...)](https://docs.taipy.io/en/latest/manuals/reference/pkg_taipy.gui/#functions) on the backend
after creating the right file

**Comment by Dr-Irv** [comment]
> I suppose the easiest way to do this would be to have a standard button with an action that would invoke [download(state, ...)](https://docs.taipy.io/en/latest/manuals/reference/pkg_taipy.gui/#functions) on the backend after creating the right file

Yes, I agree.  My colleague is going to try that now.  Then we can put the creation of the file in a context manager

**Comment by FredLL-Avaiga** [comment]
please, let us know how it goes.

**Comment by Dr-Irv** [comment]
> please, let us know how it goes.

Not well.  If we use `NamedTemporaryFile` in a context manager, then the download fails because the UI asks for the file and it is no longer there.

For right now, we are placing the file in `tempfile.gettempdir()` and using `download()`, but then on Linux, `/tmp` will start filling up with those files, and they really should be deleted once downloaded.

So I think you need a kind of action that tells python "download is complete".

**Comment by FredLL-Avaiga** [comment]
I'll look into it

**Comment by FredLL-Avaiga** [comment]
on_action is now called at the end of the download

**Comment by Dr-Irv** [comment]
> on_action is now called at the end of the download

Cool.  Will there be a version 2.4.1 that would include this?

**Comment by FredLL-Avaiga** [comment]
if you really need it :-)

**Comment by Dr-Irv** [comment]
> if you really need it :-)

It would be helpful, since we're deploying an app to production with this feature. 

Would also be useful to see if you can include a fix for #886

**Comment by FredLL-Avaiga** [comment]
#886 is also fixed in the coming 3.0

**Comment by FabienLelaquais** [comment]
Irv.
We are currently trying to fix as many problems seen in Taipy 2.x as we can for 3.0 to be delivered as shiny as possible.
At this point, 3.0 should be available un less than a month. In the meanwhile, we can create 'dev' releases you can consider as pre-releases.

Is it that crucial for you to use 2.x? Wouldn't you be willing to upgrade as soon as possible?
Backporting fixes do take a little time and potentially jeopardize the overall robustness... Just sayin'

**Comment by Dr-Irv** [comment]
We have workarounds for those issues, so we can wait for 3.0.

**Comment by FabienLelaquais** [comment]
That's great news, thanks.
We will certainly build a dev release of Taipy 3.0 (including config, core, and gui) so you can verify that workarounds are no longer needed.

**Comment by Dr-Irv** [comment]
> on_action is now called at the end of the download

@FredLL-Avaiga 

I discussed this with my colleague who has been implementing this, and we're not exactly sure how your fix would work.  Here is what we think should happen:

- If using the `file_download` control, there should be two callbacks.  One is called before the download begins, and one is called when the download is complete.  In the before-callback, we would write code to create the file to be downloaded.  In the after-callback,  we would delete the file created in the before-callback.
- If using the `download()` function, you'd need a parameter to that function to indicate the callback to use when the download was complete.  As an example, you could have your own button that has an `on_action()` callback, then that callback creates the file, then it calls `download()`, but it needs an "after-callback" to know when to delete the file.  (We are using the button/on_action/download combination for now, and just letting files accumulate in `/tmp`)

Can you indicate whether things would work as described above?

**Comment by FredLL-Avaiga** [comment]
Hi @Dr-Irv
- using the `file_download` control, we don't have a pre download action (that would mean that the control would have to wait for the backend action to execute before starting the download).
- when using the `download()` function, an action name can be passed as a parameter (on_action) (there was a bug that prevented the action to be called: fixed)
 
So I guess the best thing for your use-case is:
- use a button to create the file and call `download()` with an `on_action`

**Comment by Dr-Irv** [comment]
> So I guess the best thing for your use-case is:
> 
> * use a button to create the file and call `download()` with an `on_action`

OK, that makes sense.  I'm not sure how useful the `file_download` control would be without the ability to dynamically create the file.  I don't know if an `on_change` event could be the thing that becomes the "before-download" callback.

**Comment by FabienLelaquais** [comment]
The request for a 'pre-download' callback makes sense. It is captured in #925 to better focus on it, although is really is the initial objective of this issue, where the topic slightly moved.
This will be looked at for sure.

**Comment by FredLL-Avaiga** [comment]
#925 is in review

**Comment by FredLL-Avaiga** [comment]
Hi Dr Irv
Did you know that you can provide the content as an array of bytes.
if the size is < 50KB (configurable) it will be send as a data URL (ie no file on disk)
if not taipy will write the content to a file ...

**Comment by Dr-Irv** [comment]
> Hi Dr Irv Did you know that you can provide the content as an array of bytes. if the size is < 50KB (configurable) it will be send as a data URL (ie no file on disk) if not taipy will write the content to a file ...

Yes, but I still need to create the content dynamically when the user pushes the "file download" button.

**Comment by FabienLelaquais** [comment]
That is available in Taipy GUI 3.0... in the next few days.
You will be able to generate data on the fly, when the user presses the download button.
Examples will be provided in the doc.
